### DISCUSSION TITLE
USB pass through

### DISCUSSION THREAD

**Discussion by jwinfieldnz** [luc-github/ESP3D#799] [Questions]
I have hijacked the serial off my CH4300 (Ender 3 V2). This removes the function of the front USB port now. Is it possible to use the USB of the ESP32 to pass the data through to the serial to return the USB function? Maybe I am using the wrong UART of the ESP32, I am using the default config.

**Comment by luc-github** [comment]
you mean connect Printer usb to ESP32 USB ?

**Reply by jwinfieldnz** [reply]
No, the UART. bypass USB serial. So the ESP32 is now doing the job of the CH4300

**Reply by luc-github** [reply]
>Is it possible to use the USB of the ESP32 to pass the data through to the serial to return the USB function ?
>you mean connect Printer usb to ESP32 USB ?
>No, the UART. bypass USB serial

Sorry I do not understand your question

ESP32 USB use serial 0, GPIO1/3

ESP3D use Serial 0 by default but you can  change it

**Reply by jwinfieldnz** [reply]
Is it possible to send printer commands via the ESP32 USB to the printer UART?

**Reply by luc-github** [reply]
I am not electronician - but I do not think so

**Reply by jwinfieldnz** [reply]
Right now it can collect serial data from port 8888 and pass this onto serial a serial port. It would be good to be able to receive serial data from the USB and also pass this to the serial port. For example https://docs.arduino.cc/built-in-examples/communication/SerialPassthrough

**Reply by luc-github** [reply]
Yes it should not be hard to add this, but is was not a priority as there was no user case for it, I am lazy guy ,I do not do things if there is not need for them.
what would be the purpose of such bridge ?

**Reply by jwinfieldnz** [reply]
I love your honesty. I added this to my machine. If I have to swap over to the built-in USB it'll be a major to pull it all apart. Having a USB pass-through I lose nothing, without it I lose USB. I made a video using your software here: https://youtu.be/chc8zRZh18A and am doing a hackaday.io article. It would be good to be able to say all benefits, not losses. 
![image](https://user-images.githubusercontent.com/110659176/183404490-1f68f9e6-3356-419f-95ff-49510649c6ce.png)

**Reply by luc-github** [reply]
Ok I see - your printer has only one serial then,

I know it is not what you expect but in short term would using this command to disable the Serial of ESP3D may help
```
* Get state / Set Enable / Disable Serial Communication
[ESP900]<ENABLE/DISABLE>
```

The bridge could be definitly implemented, but will be in 3.0 which should allow this addon pretty easily 
You may have a look to it : https://github.com/luc-github/ESP3D/tree/3.0
3.0 come with lot of new features compare to 2.x
https://github.com/luc-github/ESP3D/blob/3.0/Features.md
https://github.com/luc-github/ESP3D-WEBUI/blob/3.0/Features.md

**Comment by luc-github** [comment]
I have created a feature request :https://github.com/luc-github/ESP3D/issues/800

**Reply by jwinfieldnz** [reply]
Awesome THX

**Reply by luc-github** [reply]
The serial bridge is now implemented , settings is also in Configurator now

**Reply by jwinfieldnz** [reply]
Legend, I will test it out.

**Reply by liqijian101** [reply]
I read the update about serial bridge, I think the serial bridge can not  realize the function  that he need.
The function he wanted was,
The ESP32 uses USB to connect to the USB of the printer's motherboard, which is connected to the CH340.
@jwinfieldnz @luc-github

**Reply by liqijian101** [reply]
To do this, you need to use the USB HOST CDC feature of ESP32

**Reply by luc-github** [reply]
no it is not usbhost
Printer has only one serial / usb
So instead of disconnect ESP serial each time you want to use Printer USB, 
you connect ESP Serial 1 or 2 to the Serial port of  printer ,  and only use the ESP USB which is Serial 0, the ESP do the bridge to the printer serial, so not need to disconnect ESP use printer usb

**Reply by luc-github** [reply]
the USB Host CDC is for block device like keyboard / mouse / may be Flash disk ? etc but it won't work with printers USB, that is my understanding

**Reply by liqijian101** [reply]
For example,Assuming that the MEGA2560 R 3 motherboard does not have an AUX1 port, we can only connect to the printer motherboard on the USB serial port that is connected to the CH340, whether the function of this serial bridge can be applied in this case

**Reply by luc-github** [reply]
yes a real case is the ender board : https://github.com/luc-github/ESP3D/wiki/Hardware-connection#creality-cr10-ender-3-board

**Reply by liqijian101** [reply]
This connection is very unfriendly, so we want to connect directly to the USB serial port.

**Reply by luc-github** [reply]
In that case you need a real USB host which ESP32 does not have
some solution like [node from fiber-punk](https://fiber-punk.com/) add a chip to USB32 to do the USB Host and currently it works only with few printers USB
The USB HOST support has already  been discussed already here : https://github.com/luc-github/ESP3D/issues/424

**Reply by liqijian101** [reply]
Yes, but I think Esp32 s3or s2 has USB OTG capabilities and should enable ESP32 to have similar capabilities to cancel out the extra chips

**Reply by luc-github** [reply]
again USB host otg and usb host for printer usb are different
That said if you can have working solution I would be happy to test it

**Reply by luc-github** [reply]
@liqijian101  as confirmation https://github.com/espressif/arduino-esp32/issues/7176 USB host is not supported in this way

**Reply by jwinfieldnz** [reply]
> The function he wanted was,
The ESP32 uses USB to connect to the USB of the printer's motherboard, which is connected to the CH340.

Not correct. I have connected directly to CH340 so am bypassing the printer USB. If this feature allows me to connect to the printer to replace CH340 this solution is ideal. I have not had a chance to test it yet. 

![image](https://user-images.githubusercontent.com/110659176/186580322-f14d4c0c-70d5-45ae-a19d-8f071c8a8eb3.png)

**Reply by jwinfieldnz** [reply]
Looking at configuration.h

```/* Main Serial port
* which serial ESP use to communicate to printer (ESP32 has 3 serials available, ESP8266 only 2)
* USE_SERIAL_0 //for ESP8266/32, also used by bootloader output, so consider to make it quiet
* USE_SERIAL_1 //for ESP8266/32
* USE_SERIAL_2 //for ESP32 Only
*/
//Main serial port
#define ESP_SERIAL_OUTPUT USE_SERIAL_0

/* Bridge Serial port (deprecated on esp8266 as second serial is)
* which serial ESP use to bridge to another device (ESP32 has 3 serials available, ESP8266 only 2)
* USE_SERIAL_0 //for ESP8266/32, also used by bootloader output, so consider to make it quiet
* USE_SERIAL_1 //for ESP8266/32
* USE_SERIAL_2 //for ESP32 Only\
* Comment if not used
*/
````
Would it work if the bridge port and main serial are the same?

**Reply by luc-github** [reply]
no you will get error message telling you it is not possible
